Add support for growing allocations

milroy opened the pull request:
This PR adds `grow_allocate` functionality, which attempts to allocate new resources specified by a jobspec to an existing jobid. The PR adds necessary functionality to the traverser, resource module, and REAPI, and adds testsuite tests for correct behavior.

~This PR is WIP~ pending client testing in FluxQueue, Fluence, and additional testsuite tests.

milroy committed:
policies: add grow_allocation

Problem: adding support for growing an allocation requires a new
match_op.

Add match_op_t::MATCH_GROW_ALLOCATION.

milroy committed:
traversers: support MATCH_GROW_ALLOCATION match op

Problem: adding support for growing an allocation requires a new
match_op.

Add match_op_t::MATCH_GROW_ALLOCATION to the traverser. While the new
match operation behaves identically to MATCH_ALLOCATE in the traverser,
creating a new operation is clearer in intent.

milroy committed:
modules: add support for grow_allocation

Problem: adding support for growing an allocation requires a new match
command in the resource module.

Add the grow_allocation command to the module and add error handling for
existent or nonexistent jobids.

milroy committed:
C++ REAPI: add CLI support for grow_allocation

Problem: adding support for growing an allocation requires a new match
operation in the REAPI CLI.

Add the MATCH_GROW_ALLOCATION operation to the CLI and add error
handling for nonexistent jobids.

milroy committed:
C REAPI: add CLI support for grow_allocation

Problem: adding support for growing an allocation requires a new match
operation in the C REAPI CLI.

Add the MATCH_GROW_ALLOCATION operation to the C CLI and add error
handling for nonexistent jobids.

milroy committed:
C REAPI: add module support for grow_allocation

Problem: the new MATCH_GROW_ALLOCATION isn't documented in the C REAPI
module.

Add an entry for MATCH_GROW_ALLOCATION.

milroy committed:
resource-query: add support for grow_allocation

Problem: adding support for growing an allocation requires handing the
grow_allocation command in resource-query.

Add the grow_allocation command to resource-query and add error handling
for nonexistent jobids.

milroy committed:
rq2: add support for grow_allocation

Problem: adding support for growing an allocation requires handing the
grow_allocation command in rq2.

Add the grow_allocation command to rq2 and add error handling for
nonexistent jobids.

milroy committed:
testsuite: add tests for grow_allocation

Problem: growing an allocation is untested in the sharness testsuite.

Add tests to the existing grow sharness test.

vsoch commented on resource/modules/resource_match.cpp:
```diff
@@ -1666,6 +1666,8 @@ static int run (std::shared_ptr<resource_ctx_t> &ctx,
             rc = tr.run (j, ctx->writers, match_op_t::MATCH_ALLOCATE_W_SATISFIABILITY, jobid, at);
         else if (std::string ("allocate_orelse_reserve") == cmd)
             rc = tr.run (j, ctx->writers, match_op_t::MATCH_ALLOCATE_ORELSE_RESERVE, jobid, at);
+        else if (std::string ("allocate_orelse_reserve") == cmd)
```
How would we get to this condition? Should this be like:

```suggestion
        else if (std::string ("grow_allocate") == cmd)
```

vsoch replied:
And minor suggestion to make it ` match_op_t::MATCH_GROW_ALLOCATE` just to match the naming convention of the others (action/verb "allocate" vs noun "allocation")

milroy replied:
> How would we get to this condition? Should this be like:

That's an error in my copy-paste. Fixed.

milroy replied:
> And minor suggestion to make it  match_op_t::MATCH_GROW_ALLOCATE just to match the naming convention of the others (action/verb "allocate" vs noun "allocation")

I considered both options and am leaning toward `match_op_t::MATCH_GROW_ALLOCATION` and `"grow_allocation"` given the precedent of `match_op_t::MATCH_ALLOCATE_W_SATISFIABILITY`.

vsoch commented on resource/policies/base/match_op.h:
```diff
@@ -6,6 +6,7 @@ typedef enum match_op_t {
     MATCH_ALLOCATE,
     MATCH_ALLOCATE_W_SATISFIABILITY,
     MATCH_ALLOCATE_ORELSE_RESERVE,
+    MATCH_GROW_ALLOCATION,
```
Suggestion for naming consistency:

```suggestion
    MATCH_GROW_ALLOCATE,
```
I shouldn't be reviewing this early, I've read that word less than 10 times and am getting semantic satiation. "ALLOCATE" looks like some kind of fruit name to me right now.

milroy replied:
I waffled on this and am back to preferring `MATCH_GROW_ALLOCATION`. My current thinking is that the pattern is verb_verb_{noun/verb/conjunction}, and `GROW_ALLOCATION` is clearer.

vsoch commented on resource/traversers/dfu.cpp:
```diff
@@ -99,7 +100,10 @@ int dfu_traverser_t::request_feasible (detail::jobmeta_t const &meta,
     // only the initial time matters for allocate
-    const int64_t target_time = op == match_op_t::MATCH_ALLOCATE ? meta.at : graph_end - 1;
+    const int64_t target_time =
+        (op == match_op_t::MATCH_ALLOCATE || op == match_op_t::MATCH_GROW_ALLOCATION)
```
Does this syntax say "if the statement is true (if it's allocate or grow) default to the meta.at, otherwise the graph end -1 (or is it the other way around)? I know this syntax exists in JavaScript and I always get it wrong.

vsoch replied:
And if that is the case, it's saying that if we grow or match allocate, we want the time to be the job metadata "at" which I assume is the time matched at? Just to check - for grow do we have a new at time for "at" set here in the job metadata (otherwise it would be the previous at associated with the jobid when the first allocation was done)?

milroy replied:
> for grow do we have a new at time for "at" set here

That's correct, the `at` time is relative to when the job request (grow or allocate) is submitted.

vsoch commented on resource/traversers/dfu.cpp:
```diff
@@ -118,7 +122,8 @@ int dfu_traverser_t::request_feasible (detail::jobmeta_t const &meta,
         // no chance, don't even try
-        if (op == match_op_t::MATCH_ALLOCATE_ORELSE_RESERVE || op == match_op_t::MATCH_ALLOCATE) {
+        if (op == match_op_t::MATCH_ALLOCATE_ORELSE_RESERVE || op == match_op_t::MATCH_ALLOCATE
+            || op == match_op_t::MATCH_GROW_ALLOCATION) {
```
EEEEEEEEEEEEBUSY _continues doing nodey things_ :keyboard:

vsoch commented on resource/traversers/dfu.cpp:
```diff
@@ -217,6 +222,7 @@ int dfu_traverser_t::schedule (Jobspec::Jobspec &jobspec,
             break;
         }
         case match_op_t::MATCH_ALLOCATE:
+        case match_op_t::MATCH_GROW_ALLOCATION:
```
I don't think I understand this syntax - does this mean do nothing for match allocate, or that match allocate is the same as match grow allocate? I would guess the latter since one line was just added (so we want EBUSY for both cases).

In that case, this is saying if no resources are available, just return busy. I think this answers my question about reserve - we can't reserve with an update. It's a yes / no, and right now, if we don't have it ask again later!

milroy replied:
> does this mean do nothing for match allocate

Kind of, yes. It's fall-through syntax, see: https://en.cppreference.com/w/cpp/language/switch.html or https://learn.microsoft.com/en-us/cpp/cpp/switch-statement-cpp?view=msvc-170

trws replied:
Just to be extra clear, maybe add the standard fallthrough annotation here. That way if we turn on fallthrough warnings it will not come back and bug us later.

https://en.cppreference.com/w/cpp/language/attributes/fallthrough
```suggestion
        [[fallthrough]];
        case match_op_t::MATCH_GROW_ALLOCATION:
```

vsoch commented on resource/utilities/command.cpp:
```diff
@@ -334,19 +340,29 @@ static int run_match (std::shared_ptr<resource_context_t> &ctx,
+        } else {
+            jobid = static_cast<int64_t> (std::strtoll (args[3].c_str (), NULL, 10));
+            if (!ctx->jobs.contains (jobid)) {
+                std::cerr << "ERROR: nonexistent jobid " << jobid << std::endl;
```
Here is (I think) a third check for existence - I am guessing this is done because the state might change, and better to be safe than sorry.

vsoch replied:
Oh maybe this is it?

vsoch replied:
Oh sorry, my browser didn't update the thread and I missed your comment - gotcha.

vsoch commented on t/data/resource/commands/elastic/cmds10.in:
```diff
@@ -0,0 +1,6 @@
+match allocate @TEST_SRCDIR@/data/resource/jobspecs/basics/test006.yaml
+find sched-now=allocated
```
I can better review these tests when I try to do them similarly in fluxion-go, I have trouble reviewing these files without that context.

vsoch reviewed (commented):
:hamburger:

milroy committed:
reapi: add allocation grow unit tests

Problem: growing an allocation does not have any unit testing.

Add unit tests of initialization and allocation growth via Catch2.

trws commented on resource/utilities/rq2.cpp:
```diff
@@ -382,7 +385,11 @@ int match (resource_query_t &ctx, std::vector<std::string> &args, json_t *params
+    if (subcmd != "grow_allocate") {
+        jobid = ctx.get_job_counter ();
+    } else {
+        jobid = jobid = static_cast<uint64_t> (std::strtoull (args[3].c_str (), NULL, 10));
```
jobid twice?

milroy committed:
reapi unit test: add shrink

Problem: shrinking an allocation does not have any unit testing.

Add unit tests of initialization and allocation shrink via Catch2.

milroy committed:
ion-resource: add support for growing an allocation

Problem: growing an allocation is not supported in flux ion-resource.

Add the ability to send match RPCs to grow an existing allocation.

milroy committed:
testsuite: add flux ion-resource tests

Problem: growing an allocation via flux ion-resource is untested.

Add tests to grow an existing allocation via flux ion-resource.

milroy commented on resource/modules/resource_match.cpp:
```diff
@@ -1990,9 +1992,17 @@ static void match_request_cb (flux_t *h, flux_msg_handler_t *w, const flux_msg_t
+            goto error;
+        }
+    } else {
+        if (strcmp ("grow_allocate", cmd) == 0) {
```
You're right on all four items.

> One case I'm interested to test is if it's possible to shrink down to 0 and then go back up

It should be possible to do that in `sched` but not in `core`.

milroy commented on resource/reapi/bindings/c++/reapi.hpp:
```diff
@@ -111,6 +111,8 @@ class reapi_t {
      *                   MATCH_ALLOCATE_W_SATISFIABILITY: try to allocate and run
      *                   satisfiability check if resources are not available.
+     *                   MATCH_GROW_ALLOCATION: try to grow an existing allocation
+     *                   now and fail if resources aren't available.
```
> Would that work here?

No, that won't work with these changes. There's a lot more machinery needed to detect cases where the reservation wouldn't start before the current allocation ends, or the reservation time changes and becomes invalid.

milroy replied:
Which makes me realize that even with `MATCH_GROW_ALLOCATION` there needs to be some way to ensure that the start time (`at`) of the request isn't invalid or after the current allocation completes.

trws commented:
I'm good with `MATCH_GROW_ALLOCATION` as long as you are.

As to the timing, I think we need to have something in free that does a cancel on any in-flight grow requests.  Speaking of that actually, I haven't searched for it but this fails if the job matching its ID is in a state other than "allocated/running" right? Stacking grows on a pending job sounds like a recipe for strange, strange cases.
